src/server/lib/okta/authentication.ts

```diff
@@ -0,0 +1,163 @@
+interface AuthenticationTransaction {
+  _embedded?: AuthenticationEmbeddedUser;
+  _links?: AuthenticationLinks;
+  expiresAt?: string;
+  factorResult?: AuthenticationFactorResult;
+  sessionToken?: string;
+  stateToken?: string;
+  status?: AuthenticationTransactionState;
+}
+
+// https://developer.okta.com/docs/reference/api/authn/#request-parameters-for-primary-authentication
+interface AuthenticationRequestParameters {
+  audience?: string;
+  context?: {
+    deviceToken?: string;
+  };
+  options?: {
+    multiOptionalFactorEnroll?: boolean;
+    warnBeforePasswordExpired?: boolean;
+  };
+  password?: string;
+  token?: string;
+  username?: string;
+}
+
+const { oktaDomain } = getConfiguration();
+
+const API_ROUTE = '/api/v1/authn';
+
+const handleResponseFailure = async (response: Response) => {
+  try {
+    const parsed = (await response.json()) as OktaError;
+    switch (parsed.errorCode) {
+      // errors from https://developer.okta.com/docs/reference/api/authn/#response-parameters
+      case 'E0000004':
+        throw {
+          status: response.status,
+          statusText: response.statusText,
+          message: OktaAuthenticateErrors.AUTHENTICATION_FAILED,
+        };
+      case 'E0000047':
+        throw {
+          status: response.status,
+          statusText: response.statusText,
+          message: SignInErrors.GENERIC,
+        };
+      default:
+        throw {
+          status: response.status,
+          statusText: response.statusText,
+          message: SignInErrors.GENERIC,
+        };
+    }
+  } catch (error) {
+    throw error;
+  }
+};
+
+const handleResponseSuccess = async (
+  response: Response,
+): Promise<AuthenticationTransaction> => {
+  try {
+    return (await response.json()) as AuthenticationTransaction;
+  } catch (error) {
+    throw new Error(`Error decoding JSON response: ${error}`);
+  }
+};
+
+// okta authentication endpoint
+// gateway (profile) is set up as a public application as it is
+// accessible by anyone
+// https://developer.okta.com/docs/reference/api/authn/#request-example-for-primary-authentication-with-public-application
+export const authenticate = async (email: string, password: string) => {
+  const request: AuthenticationRequestParameters = {
+    username: email,
+    password,
+    options: {
+      multiOptionalFactorEnroll: false,
+      warnBeforePasswordExpired: false,
+    },
+  };
+
+  const body = JSON.stringify(request);
+
+  const requestOptions: RequestInit = {
+    method: 'POST',
+    headers: {
+      Accept: 'application/json',
+      'Content-Type': 'application/json',
+    },
+    body,
+    agent: getAgent,
+  };
+
+  try {
+    const response = await fetch(`${oktaDomain}${API_ROUTE}`, requestOptions);
+    if (!response.ok) {
+      return await handleResponseFailure(response);
+    } else {
+      return await handleResponseSuccess(response);
+    }
+  } catch (error) {
+    throw error;
```

Comment:
the try catch is unnecessary if the error is thrown

Reply:
Ah there isn't, it's a hangover from when I implemented it and was testing errors, try catch it's required here. Good spot.